rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,253 @@
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+MFA device "solokey" added.
+
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Last used
+----------------   ----   ------------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    c8fbb126-3a29-4a9c-bfe9-ebaed31d8585   Wed 16 Dec 2020 02:00:13 PM PST
+solokey            U2F    87d4fb03-012c-451c-ab0d-5f2a681c119a   Wed 16 Dec 2020 02:05:46 PM PST
+
+# remove by name
+$ tsh mfa rm yubikey
+Tap any *registered* security key... <tap>
+MFA device "yubikey" removed.
+
+# remove by ID
+$ tsh mfa rm fa004bf4-acc7-435d-8965-5f5a0a4552e8
+Tap any *registered* security key... <tap>
+MFA device "android OTP" removed.
+
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Last used
+----------------   ----   ------------------------------------   -------------------------------
+solokey            U2F    87d4fb03-012c-451c-ab0d-5f2a681c119a   Wed 16 Dec 2020 02:05:46 PM PST
+
+# If 2FA is optional:
+$ tsh mfa rm solokey
+You are about to remove the only remaining MFA device.
+This will disable MFA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh mfa rm solokey
+Can't remove the only remaining MFA device.
+Please add a replacement MFA device first using "tsh mfa add".
+```
+
+#### web UI
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
+- buttons to remove any enrolled device (except for the last one)
+
+Web UI details, wireframes and implementation will be added later, when we have
+the capacity to do it. Initially, 2FA management is CLI-only.
+
+#### bootstrap
+
+Initially, a user account doesn't have a 2FA device. Depending on [cluster
+configuration](#configuration), use of 2FA devices might be required.
+
+If 2FA is required, a user is required to enroll a device during account
+creation (for local users) or first login (for SSO users).
+
+If 2FA is optional, a user can create an account and login without 2FA. They
+can then add 2FA devices as described above. If an existing user has at least
+one 2FA device registered, it's required during login.
+
+### configuration
+
+The current 2FA configuration in Teleport only applies to local users and is
+always required. We need to allow this to be optional (for users to migrate)
+and allow to mix different 2FA methods.
+
+Existing configuration options must be backwards-compatible - no change in
+behavior unless config values are changed.
+
+New values for `auth_service.authentication.second_factor` for this:
+- `off` (existing) - no 2FA can be enrolled
+- `otp` (existing) - only OTP can be enrolled and is required for all local
+  users
+- `u2f` (existing) - only U2F can be enrolled and is required for all local
```

Comment:
Enforcement for U2F only is particularly important as I would expect many customers to use this. In practice, U2F is the only protection agains phishing.